### ISSUE TITLE
Inline authorization handler

### ISSUE THREAD

**Issue by cts-tradeit** [ChilliCream/hotchocolate#5270]
### Is your feature request related to a problem?

You may use ASP NET Core's authorization handlers to restrict access to an object type.
```
class CarType: ObjectType<Car> {
    Configure(IObjectTypeDescriptor<Car> descriptor) {
          descriptor.Authorize("CanDrive").Authorize("AtLeast18");
    }
}
class BoozeType: ObjectType<Booze> {
    Configure(IObjectTypeDescriptor<Booze> descriptor) {
          descriptor.Authorize("AtLeast18");
    }
}
class BikeType: ObjectType<Bike> {
    Configure(IObjectTypeDescriptor<Bike> descriptor) {
          descriptor.Authorize("CanDrive");
    }
}
```
This allows you to create reusable policies as long as they do not work with the underlying object (policies above may assert against instance of User but cannot assert against instance of Car otherwise cannot be reused).

When your security depends mostly on entity related data to authorize the access you may end up with dozens and hundreds of different policies.
```
options.AddPolicy("ReadContract", ...);
options.AddPolicy("ReadClient", ...);
options.AddPolicy("ReadTrade", ...);
...

services.AddSingleton<IAuthorizationHandler, ReadContractHandler>();
services.AddSingleton<IAuthorizationHandler, ReadClientHandler>();
services.AddSingleton<IAuthorizationHandler, ReadTradeHandler>();
...

public class ReadContractHandler: AuthorizationHandler<Requirement, IResolverContext> {
    Task HandleRequirementAsync(AuthorizationHandlerContext context, Requirement requirement,  IResolverContext resource)
    {
         Contract contract = resource.Parent<Contract>();
         var accessibleContracts = resource.DataLoader<Loader>().LoadAsync(userId /*ignore dec. err.*/);
        if (accessibleContracts.Contains(contract.Id))
            context.Succeed(requirement);
        else 
            context.Fail();
    }
} 
```
Each of these handlers has to query database because readability is based through DB relations and column values, rather than static roles or traits of current user.

Basically every ObjectType has own policy which 1 requirement, 1 handler, 1 policy name. 

### The solution you'd like

Ability to specify additional rules through delegates can remove this overhead for single use rules:
```
class CarType: ObjectType<Car> {
    Configure(IObjectTypeDescriptor<Car> descriptor) {
          descriptor.Authorize(ctx => { 
              Car car = ctx.Parent<Car>();
              bool carRegistered = ctx.DataLoader<StuffLoader>().LoadAsync(/*...*/).Contains(car.VIN);
              if (!carRegistered) ctx.Fail();
          });
    }
}
```
This does not have to replace use of ASP.NET Core AuthorizationHandlers. Methods like Authorize(string policy) might just add a handler that will invoke given policy.

### Product

Hot Chocolate

**Comment by cts-tradeit** [comment]
This issue is also addressed by solution proposed in #5273 using `IAuthorizationService`;

**Comment by ademchenko** [comment]
A very similar issue has been started discussing more than one year ago: #3456 . The authorization mechanisms of Hotchocolate are really not quite powerful.

**Comment by kolpav** [comment]
I encountered exactly the same pain points. For me the biggest one was at the time I was still learning learning .net auth itself so I always assumed I am doing something wrong and ended up with a lot of ugly code having to do multiple additional requests on top of regular queries I could describe like 2 - 3 auth scenarios which I think are impossible to solve now feel free to ping me if that would be useful to you.

If there is going to be rework of authorization it would be nice if it would focus on resource based authorization from the start.

**Comment by kolpav** [comment]
Below is example of authorization I think would be hard to do now its different from what I had to solve but I think it includes requirements/restrictions which might help to uncover rough edges of current implementation.

@michaelstaib It would really help me if you could take a look and point me to direction of some existing functionality which would make such auth possible even in yet to be release code just keywords to search for are more than enough, thank you.

## Rules

- author is meant as **who** created the entity
- owner as in **relationship** with other types
- not having access means there is no way for you do know if entity exists or not

**Issue**
- `PRIVATE`
  - Accessible only if you are the author
- `PUBLIC`
  - No rules can be part of other queries
- `ARCHIVED`
  - Only if you know the `Id`

**Comment**
- Owner is `PRIVATE`
  - Accessible only by the author
- Owner is `PUBLIC`
  - No rules
  - Can be part of other queries where its not being owned by `Issue`
- Owner is `ARCHIVED`
  - Accessible to anyone but if you are not the author 
 `CreatedAt` will have just a year part with everything else being `0`


## Schema

The schema is bit weird so don't look into it that much :)

- Assume the schema is 1:1 to database schema
- You cannot edit db schema
- You can edit gql schema

```gql
enum State { PRIVATE, PUBLIC, ARCHIVED }

type Issue { id: ID!, State: State!, Comments: [Comment]!, CreatedBy: User! }
type Comment { id: ID!, CreatedAt: Date! }

# NOTE: There is no query accepting just an `Id`! 
type Query {
  getIssue(input: IssueFilterInput!): Issue
  getIssues(input: IssueFilterInput!): [Issue]!
  getComment(input: CommentFilterInput!): Comment
  getComments(input: CommentFilterInput!): [Comment]!
}
```

_Might contain syntax errors or typos_

**Comment by cts-tradeit** [comment]
For rule `PRIVATE` i'd suggest using object type policies as they are called anytime given type appears in the query (no matter where in the schecma). More: https://chillicream.com/docs/hotchocolate/security/authorization

```csharp
// Policy
class IssueRequirement: IAuthorizationRequirement { }
class IssueHandler: AuthorizationHandler<IssueRequirement, IOperationContext> {
    protected override HandleRequirementAsync(AuthorizationHandlerContext ctx1, IssueRequirement req, IResolverContext ctx2) 
    {
        Issue issue = ctx2.Parent<Issue>();
        int userId = ctx2.Service<IHttpContextAccessor>().HttpContext.User.Id();

        // If u need load something from DB use dataloader to merge concurent requests into one
        // var relatedData = ctx2.DataLoader<SomeLoader>().LoadAsync(...);

        if (issue.State == States.Private && userId != issue.CreatedById)    
            return ctx1.Fail();
        return ctx1.Succeed(req);
    }
}
// Startup
services.AddAuthorizationHandler<IssueRequirement>();
services.AddAuthorization(opts => {
    opts.AddPolicy("ReadIssue", o => o.AddRequirement<IssueRequirement>());
});
// Object type (or use attribute on entity class [Authorize("ReadIssue")])
class IssueObjectType: ObjectType<Issue> {
    protected override Configure(IObjectTypeDescriptor<Issue> descriptor) {
        descriptor.Authorize("ReadIssue");
    }
}
```

For `ARCHIEVED` it would make sense to create another policy but place it on the resolver method instead of object type as it asserts against resolver's parameter `input` rather than entity itself. However, you cannot access result even if you setup the policy to be applied after the method is run using `ApplyPolicy.AfterResolver`. So I'd suggest creating custom field middleware. More: https://chillicream.com/docs/hotchocolate/execution-engine/field-middleware

```csharp
class CheckIssueFilterAttribute: ObjectFieldDescriptorAttribute {
    public override void OnConfigure(IDescriptorContext context, IObjectFieldDescriptor descriptor, MemberInfo member)
    {
        descriptor.Use(next => async ctx => {

            // Run next middleware (ctx.Result is accessible only after)
            await next(ctx);

            // name can be passed in ctor of the attribute
            IssueFilterInput filter = ctx.Argument<IssueFilterInput>("input");
            object result = ctx.Result;

            // use data loader to query additional data
            // var related = ctx.DataLoader<FooLoader>().LoadAsync(...);

            if (result is Issue issue)
                return Inspect(issue, filter);
            if (result is IEnumerable<Issue> issues) {
                foreach(Issue issue in issues)
                    Inspect(issue, filter);
            }
            throw new Exception("Result was not Issue or enumerable of Issue");
        });
    }

    private void Inspect(Issue issue, IssueFilterInput filter) {
        if (issue.State == States.Archived && filter.Id is null)
            throw new InvalidOperationException("Archived issue can be queried only by id.");
    }
}

// If you declare Queries as object type, do not create an attirbute but use .Use() method in Configure method directly
class Queries {
    [CheckIssueFilterAttribute]
    public Task<Issue> GetIssue(IssueFilterInput input) {
        /*...*/
    }
    [CheckIssueFilterAttribute]
    public Task<IEnumerable<Issue>> GetIssues(IssueFilterInput input) {
        /*...*/
    }
}
```

For `ARCHIVED` as it's not a policy but rather a transformation, I'd suggest creating custom resolver method (and override the resolver infered from the propety). You can theoreticly change value of the property inside the middleware, but it would take affect only where the attribute/middleware is placed and not globally as trait of the object type. More: https://chillicream.com/docs/hotchocolate/defining-a-schema/extending-types

```csharp
// If u define ObjectType class, you can use .Field(_ => _.CreatedAt).Resolver(ctx => ...);
// Attributed parameters are accessible by ctx.Parent<Comment>(), ctx.DataLoader<Loader>() etc...
// Watchout CancellationToken is in ctx.RequestAborted
[ExtendObjectType(typeof(Comment))]
class CommentExtension {
    // Set explicitly Date, otherwise DateTime will be inferred
    [GraphQLType(typeof(NonNullType<DateType>))]
    public Task<DateTime> GetCreatedAt([Parent] Comment comment, [DataLoader] EntityByIdLoader<Issue> loader, 
        [Service] IHttpContextAccessor accessor, CancellationToken token) {
        Issue issue = await loader.LoadAsync(comment.IssueId);
        int userId = ctx2.Service<IHttpContextAccessor>().HttpContext.User.Id();

        if (issue.State != States.Archived) return comment.CreatedAt;
        if (issue.CreatedById == userId) return comment.CreatedAt;

        // Can't recall how to do this, so just google it
        return MakeAllMinButYear(comment.CreateAt);
    }
}
```

**Comment by michaelstaib** [comment]
For Version 13 we have abstracted authorization ... you can now implement a custom 'IAuthorizationHandler' and implement your own thing in there. 

From our perspective we do not want to implement such a thing on our own.

Instead we support out of the box OPA and the ASP.NET core auth model. Additional as mentioned you can now build your own.
